plugins/mediatek-scaler/fu-mediatek-scaler-device.c

```diff
@@ -0,0 +1,971 @@
+/*
+ * Copyright (C) 2023 Crag Wang <crag.wang@dell.com>
+ * Copyright (C) 2023 Mediatek Inc.
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <glib/gstdio.h>
+#include <linux/i2c-dev.h>
+
+#include "fu-mediatek-scaler-common.h"
+#include "fu-mediatek-scaler-device.h"
+#include "fu-mediatek-scaler-firmware.h"
+#include "fu-mediatek-scaler-struct.h"
+
+#define DDC_DATA_LEN_DFT	0x80
+#define DDC_DATA_FRAGEMENT_SIZE 0x0B   /* 11 bytes for each DDC write */
+#define DDC_DATA_PAGE_SIZE	0x1000 /* 4K bytes for each block page */
+#define DDC_RW_MAX_RETRY_CNT	10
+
+/* timeout duration in ms to i2c-dev operation */
+#define FU_MEDIATEK_SCALER_DEVICE_IOCTL_TIMEOUT 5000
+
+/* delay time before a ddc read or write */
+#define FU_MEDIATEK_SCALER_DDC_MSG_DELAY_MS 50
+
+/* interval in ms between the poll to check device status */
+#define FU_MEDIATEK_SCALER_DEVICE_POLL_INTERVAL 1000
+
+/* firmware payload size */
+#define FU_MEDIATEK_SCALER_FW_SIZE_MAX 0x100000
+
+struct _FuMediatekScalerDevice {
+	FuUdevDevice parent_instance;
+	FuUdevDevice *i2c_dev;
+	gchar *i2c_dev_name;
+	gchar *dp_aux_dev_name;
+};
+
+G_DEFINE_TYPE(FuMediatekScalerDevice, fu_mediatek_scaler_device, FU_TYPE_UDEV_DEVICE)
+
+static gboolean
+fu_mediatek_scaler_ensure_device_address(FuMediatekScalerDevice *self,
+					 guint8 address,
+					 GError **error)
+{
+	if (!fu_udev_device_ioctl(self->i2c_dev,
+				  I2C_SLAVE,
+				  (guint8 *)(guintptr)address,
```

Comment:
I'm a bit confused what the cast to guintptr is achieving.

Reply:
It creates a pointer to the `address` integer value, without the cast the compiler complains the data type inconsistence. Any better way we can pass the value?

Reply:
Nah, it's fine -- just as long as it's required :)